source/common/upstream/conn_pool_map_impl.h

```diff
@@ -95,12 +97,19 @@ bool ConnPoolMap<KEY_TYPE, POOL_TYPE>::freeOnePool() {
 
   if (pool_iter != active_pools_.end()) {
     // We found one. Free it up, and let the caller know.
+    thread_local_dispatcher_.deferredDelete(std::move(pool_iter->second));
```

Comment:
Maybe I'm missing something but this seems new and not related to the resource tracking?

Reply:
Yeah, I realized I was being inconsistent with how the "clear" works when doing this change. I figured I could do the cleanup here, but I can understand if you'd prefer it in a different PR. I can back it out if you'd like.

Reply:
To be clear, I didn't want to delete the connection pool directly while trying to allocate a new one. I figured it made more sense to use the deferredDelete mechanism. I may have misunderstood its purpose, though.

Reply:
In general we only used deferred delete when there are stack unwinding implications for doing a direct delete. If that's not the case, I would continue during a direct delete. (I would have to analyze this case to understand whether there are potential stack unwind issue but maybe you already know.)

Reply:
Okay. I remember why it was necessary in the `clear()` case. We potentially invoke `clear()` from within a connection pool callback. We don't want to delete the invoker from there -- so we defer it.

Here, we're deleteing the connection pool when allocating a new one. We will not get to this case while the connection pool being deleted is on the stack, so I do not think we need it. I'll remove the delete.

Reply:
Even that isn't great... Running into a bunch of corner cases and race conditions. *sigh*.